### PR TITLE
Finishes adding all the codepen embed options available

### PR DESCRIPTION
**Pull request by JaneOri**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [x] Documentation Update

## Description

Adds the following options to codepen liquid tag embeds:
1) allow /pen/preview/ paths (which become /embed/preview/) For #12130 
2) allow height param for #5346
3) allow theme-id for codepen pro users (this is the thing I wanted, just decided to take care of the rest while I was in there)
4) and for the sake of completion, add the last embed option codepen has too: editable=true (also for codepen pro users)

## Related Tickets & Documents

#12130
#5346

## QA Instructions, Screenshots, Recordings

Several unit tests are included

### UI accessibility concerns?

N/A

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] Are there any post deployment tasks we need to perform?

Celebration and joy

## [optional] What gif best describes this PR or how it makes you feel?

https://i.imgur.com/0JE4B1J.mp4

### CODE REVIEW

**Comment by JaneOri** [comment]
This was my first time even looking at a ruby file, so if there's something obvious I need to do for that failing CI build, please do point it out and I'll be happy to learn and fix it asap:
https://app.travis-ci.com/github/forem/forem/jobs/555584075#L1002
-->
https://github.com/forem/forem/pull/16102/files#diff-0d3a28fa9217a545d1edc51352c8f20f282b330ef73c9602398e9d10418e37f6R34

Thank you! :)

**Review by citizen428 — APPROVED** [review]

**Review by citizen428 — CHANGES_REQUESTED** [review]
Sorry, I accidentally hit "Approve" 🤦 

The reason the specs fail is that there's a syntax error:

```
SyntaxError:
  /home/travis/build/forem/forem/spec/liquid_tags/codepen_tag_spec.rb:34: syntax error, unexpected local variable or method, expecting `do' or '{' or '('
          )
     ^~
  /home/travis/build/forem/forem/spec/liquid_tags/codepen_tag_spec.rb:34: syntax error, unexpected ')', expecting `end'
          )
              ^
  /home/travis/build/forem/forem/spec/liquid_tags/codepen_tag_spec.rb:43: syntax error, unexpected local variable or method, expecting `do' or '{' or '('
          )
     ^~
  /home/travis/build/forem/forem/spec/liquid_tags/codepen_tag_spec.rb:43: syntax error, unexpected ')', expecting `end'
          )
              ^
  /home/travis/build/forem/forem/spec/liquid_tags/codepen_tag_spec.rb:103: syntax error, unexpected local variable or method, expecting `do' or '{' or '('
          )
     ^~
  /home/travis/build/forem/forem/spec/liquid_tags/codepen_tag_spec.rb:103: syntax error, unexpected ')', expecting `end'
          )
              ^
```

**Comment by JaneOri** [comment]
@citizen428 The reported syntax error is pointing to code that I didn't change, inside an old test, only on "Node 0":
image

I've never opened a Ruby file before this PR and I do not know what the problem actually is / how to fix it.

If you could identify the actual problem and let me know what to do, I'd be happy to update the PR :)

**Comment by citizen428** [comment]
The syntax error is in `spec/liquid_tags/codepen_tag_spec.rb` and according to the diff you changed the entire file 🤔  Let me have a look at get back to you in a bit.

**Comment by citizen428** [comment]
@JaneOri I'm not quite sure what the problem with your version of `spec/liquid_tags/codepen_tag_spec.rb` was, so I checked it out from `main` again and then manually re-added your changes, no more errors.

**Review by citizen428 — APPROVED** [review]

**Comment by JaneOri** [comment]
heck yeah, awesome. Might have been windows line endings? 😬

Thank you!

**Comment by citizen428** [comment]
Ah, yes, that could have been in. FWIW, you can configure `git` to always use Unix line-endings, see the documentation for `core.autocrlf` [here](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration). For example, I have mine set to `input` (`git config --global core.autocrlf input`) which behaves as follows:

> This setup should leave you with CRLF endings in Windows checkouts, but LF endings on macOS and Linux systems and in the repository.



.

**Review by jeremyf — APPROVED** [review]

**Comment by jeremyf** [comment]
@JaneOri looks great, I want to give @msarit a chance to review as well.  She's been spending a lot of time in the liquid tags.

I've added this to my watch list to merge on Tuesday UTC 20:00.

**Comment by nickytonline** [comment]
> > Might have been windows line endings?
> 
> Ah, yes, that could have been in. FWIW, you can configure `git` to always use Unix line-endings, see the documentation for `core.autocrlf` [here](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration). For example, I have mine set to `input` (`git config --global core.autocrlf input`) which behaves as follows:
> 
> > This setup should leave you with CRLF endings in Windows checkouts, but LF endings on macOS and Linux systems and in the repository.

@citizen428, we can add a config for this to the repo no? It's been a while, but I remember doing this on an old work project to avoid this.

**Comment by citizen428** [comment]
@nickytonline Sure, we can. Just change the `--global` to `--local` and execute the `git` command in a Forem checkout.

File: app/liquid_tags/codepen_tag.rb [hunk 1]
```diff
@@ -1,20 +1,22 @@
 class CodepenTag < LiquidTagBase
   PARTIAL = "liquids/codepen".freeze
   REGISTRY_REGEXP =
-    %r{\A(http|https)://(codepen\.io|codepen\.io/team)/[a-zA-Z0-9_\-]{1,30}/pen/([a-zA-Z0-9]{5,32})/{0,1}\z}
+    %r{\A(http|https)://(codepen\.io|codepen\.io/team)/[a-zA-Z0-9_\-]{1,30}/(pen|pen/preview)/([a-zA-Z0-9]{5,32})/{0,1}\z}
```
Comment (msarit) [inline app/liquid_tags/codepen_tag.rb:4]: Fringe case, but some users may grab their link from the Codepen "Embed" modal. This link is of the format:
```ruby
https://codepen.io/msarit/embed/eYJZdmp?default-tab=html%2Cresult&editable=true
```

I think we should add `embed` in the third capture group `(pen|pen/preview)`.

Also I am not sure if it is worth it in this PR to handle possible url params like we see in the example link above. Perhaps that can be a separate PR. 😅

File: app/liquid_tags/codepen_tag.rb [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
 
   def initialize(_tag_name, link, _parse_context)
     super
     @link = parse_link(link)
```
Comment (msarit) [inline app/liquid_tags/codepen_tag.rb:8]: To handle any possible escaped characters in the link (eg `%2C`), I think we should apply `CGI.unescape_html` (similar to how it's done [here](https://github.com/forem/forem/blob/71dcc7d53f5e069debc8100edcf85fb9d5b48ba8/app/liquid_tags/codesandbox_tag.rb#L13)).

File: app/views/pages/_editor_guide_text.html.erb [hunk 3]
```diff
@@ -283,6 +283,36 @@
...
+        <pre>{% codepen https://codepen.io/propjockey/pen/dyVMgBg theme-id=40148 %}</pre>
+      </dd>
+
+      <dt><strong>editable=true</strong></dt>
+      <dd>
+        <p>Add editable=true parameter to your CodePen embed tag. Editable Embeds require more resources than non-editable Embeds. The pen must be from a codepen pro user to have an effect.</p>
+        <pre>{% codepen https://codepen.io/propjockey/pen/dyVMgBg editable=true %}</pre>
+      </dd>
+
+      <dt><strong>multiple-params/strong></dt>
```
Comment (msarit) [inline app/views/pages/_editor_guide_text.html.erb:311]: Didn't close your `<strong>` tag here.

Screen Shot 2022-01-18 at 11 50 55 AM

File: app/views/pages/_editor_guide_text.html.erb [hunk 4]
```diff
@@ -283,6 +283,36 @@
         <p>Add default-tab parameter to your CodePen embed tag. Defaults to <i>result</i>.</p>
         <pre>{% codepen https://codepen.io/twhite96/pen/XKqrJX default-tab=js,result %}</pre>
       </dd>
+
+      <dt><strong>preview</strong></dt>
```
Comment (msarit) [inline app/views/pages/_editor_guide_text.html.erb:287]: These "sub-headings" are not bold by default in the Liquid Tags docs; they are highlighted instead. Please change.

Screen Shot 2022-01-18 at 11 45 40 AM

File: app/views/pages/_editor_liquid_help.html.erb [hunk 5]
```diff
@@ -145,6 +145,36 @@
         Add default-tab parameter to your CodePen embed tag. Default to <i>result</i><br>
         <code>{% codepen https://codepen.io/twhite96/pen/XKqrJX default-tab=js,result %}</code>
       </dd>
+
+      <dt><strong>preview</strong></dt>
```
Comment (msarit) [inline app/views/pages/_editor_liquid_help.html.erb:149]: Same comment as above; please change.

File: app/views/pages/_editor_liquid_help.html.erb [hunk 6]
```diff
@@ -145,6 +145,36 @@
...
+        <pre>{% codepen https://codepen.io/propjockey/pen/dyVMgBg theme-id=40148 %}</pre>
+      </dd>
+
+      <dt><strong>editable=true</strong></dt>
+      <dd>
+        <p>Add editable=true parameter to your CodePen embed tag. Editable Embeds require more resources than non-editable Embeds. The pen must be from a codepen pro user to have an effect.</p>
+        <pre>{% codepen https://codepen.io/propjockey/pen/dyVMgBg editable=true %}</pre>
+      </dd>
+
+      <dt><strong>multiple-params/strong></dt>
```
Comment (msarit) [inline app/views/pages/_editor_liquid_help.html.erb:173]: Same comment as above; please change.

File: spec/liquid_tags/codepen_tag_spec.rb [hunk 7]
```diff
@@ -7,6 +7,16 @@
     let(:codepen_team_private_link) { "https://codepen.io/team/codepen/pen/fb02c34281cb08966ec44b4e1ae22bc3" }
     let(:codepen_team_link) { "https://codepen.io/team/keyframers/pen/ZMRMEw" }
     let(:codepen_link_with_default_tab) { "https://codepen.io/twhite96/pen/XKqrJX default-tab=js,result" }
+    let(:codepen_link_with_theme_id) { "https://codepen.io/propjockey/pen/dyVMgBg theme-id=40148" }
```
Comment (msarit) [inline spec/liquid_tags/codepen_tag_spec.rb:10]: Very robust specs, great job! 🎉

**Review by msarit — CHANGES_REQUESTED** [review]
@JaneOri You've done exceptional work here - thank you so much!

I've suggested a few minor edits.

**Comment by JaneOri** [comment]
I'll address each item this evening after work, thank you!

File: app/liquid_tags/codepen_tag.rb (hunk 1 above)
Comment (JaneOri) [reply app/liquid_tags/codepen_tag.rb:4]: Great call, updating it to `/(pen|embed)(/preview)?/` and I'll add tests to cover those too

parsing the params would be a nice improvement for ease of use... agreed with doing that another time though 😅

Fortunately with this suggested change, copying the full embed url from codepen and manually replacing `?` and `&` with spaces is the only translation the author needs to do for it to work in the liquid tag. So it's not toooo much overhead now 💜

File: app/views/pages/_editor_guide_text.html.erb (hunk 4 above)
Comment (JaneOri) [reply app/views/pages/_editor_guide_text.html.erb:287]: Ah, I didn't notice that difference. `app/views/pages/_editor_guide_text.html.erb` is using `<strong>` on all of them,  `app/views/pages/_editor_liquid_help.html.erb` (in the screenshot) is using `<code>` though instead so my strong's are wrong there.

I'll update it in `app/views/pages/_editor_liquid_help.html.erb` so both files are consistent within themselves.

Happy to change it however you'd like, but I suspect this is what you'd want to see until these files are DRY'd up - lmk!

File: app/liquid_tags/codepen_tag.rb (hunk 2 above)
Comment (JaneOri) [reply app/liquid_tags/codepen_tag.rb:9]: added this here: https://github.com/forem/forem/pull/16102/files#diff-1bbc1700483e366b249f440ff13cc97f8a82a94a334d12693d53ad91b4ab2e65R8

**Comment by JaneOri** [comment]
@msarit Good to go? :D

**Review by msarit — APPROVED** [review]

**Comment by citizen428** [comment]
Thanks @JaneOri!
